app/models/reports/schema_v2.rb

```diff
@@ -0,0 +1,286 @@
+module Reports
+  class SchemaV2
+    include BustCache
+    include Memery
+
+    attr_reader :control_rate_query_v2
+    attr_reader :periods
+    attr_reader :period_hash
+    attr_reader :period_type
+    attr_reader :regions
+
+    delegate :cache, :logger, to: Rails
+
+    def initialize(regions, periods:)
+      @regions = regions
+      @periods = periods
+      @period_type = @periods.first.type
+      @control_rate_query_v2 = ControlRateQueryV2.new
+      @period_hash = lambda { |month_date, count| [Period.month(month_date), count] }
+    end
+
+    # Returns the earliest patient record for a Region from either assigned or registered patients. Note that this *ignores*
+    # the periods that are passed in for the Repository - this is the true 'earliest report date' for a Region.
+    memoize def earliest_patient_recorded_at
+      region_entries = regions.map { |region| RegionEntry.new(region, __method__) }
+      cached_results = cache.fetch_multi(*region_entries, force: bust_cache?) { |region_entry|
+        earliest_patient_data_query_v2(region_entry.region)
+      }
+      cached_results.each_with_object({}) { |(region_entry, time), results| results[region_entry.slug] = time }
+    end
+
+    memoize def earliest_patient_recorded_at_period
+      earliest_patient_recorded_at.each_with_object({}) { |(slug, time), hsh| hsh[slug] = Period.new(value: time, type: @period_type) if time }
+    end
+
+    def earliest_patient_data_query_v2(region)
+      FacilityState.for_region(region).where("cumulative_registrations > 0 OR cumulative_assigned_patients > 0")
```

Comment:
Finding the 'birth date' of a region.

Reply:
Nice! I am trying to understand why it is important to treat facilities without any data differently. I.e, why can't we just show 0s in the graphs or tables, perhaps?

Reply:
Also, can we inline this query?

Reply:
We limit the number of months shown on our graphs based on a region's birth date. If you've only got 5 months of data, no need to show 24 datapoints crowded together, we should 5 data points spaced apart (visually).

Reply:
yup, what @hari said 😄